sys/net/gnrc/conn/tcp/gnrc_conn_tcp.c

```diff
@@ -0,0 +1,560 @@
+        queue[i].status |= GNRC_TCP_STATUS_PASSIVE;
+
+        /* If the specifing conn struct, allows connection from an unspecified address mark connections */
+#ifdef MODULE_GNRC_IPV6
+        if(ipv6_addr_is_unspecified((ipv6_addr_t *)&(conn->local_addr))){
+            queue[i].status |= GNRC_TCP_STATUS_ALLOW_ANY_ADDR;
+        }
+#endif
+
+        /* Open Connections */
+        if (_fsm(&queue[i], CALL_OPEN, NULL, NULL, 0) == -ENOMEM) {
+            DEBUG("gnrc_conn_tcp : conn_tcp_listen() : Out of receive buffers, increase GNRC_TCP_RCV_BUFFERS\n");
+            assert(0);
+        }
+    }
+    return 0;
+}
+
+int conn_tcp_accept(conn_tcp_t *conn, conn_tcp_t **out_conn)
+{
+    msg_t msg;
+    conn_tcp_t *iter = NULL;
+
+    assert(conn->l4_type == GNRC_NETTYPE_TCP);
+
+    /* Guard: Specifing struct must be passive(listen was called on it). */
+    if (!(conn->status & GNRC_TCP_STATUS_PASSIVE)) {
+        return -EOPNOTSUPP;
+    }
+
+    /* Check if there is a unaccepted connection listening on the fitting port */
+    mutex_lock(&_list_conn_tcp_lock);
+    iter = _list_conn_tcp_head;
+    while (iter) {
+        if (conn->local_port == iter->local_port && !(iter->status & GNRC_TCP_STATUS_ACCEPTED)) {
+            iter->status |= GNRC_TCP_STATUS_ACCEPTED;
+            break;
+        }
+        iter = iter->next;
+    }
+    mutex_unlock(&_list_conn_tcp_lock);
+
+    /* All connections are currently handled. Nothing to do here */
+    if (iter == NULL) {
+        return -EAGAIN;
+    }
+
+    /* Take responsability for this connection */
+    *out_conn = iter;
+    iter->owner = thread_getpid();
+    msg_init_queue(iter->msg_queue, GNRC_TCP_MSG_QUEUE_SIZE);
+
+    /* Purge message queue */
+    while(msg_try_receive(&msg) >= 0){
+    }
+
+    /* Wait until connection establishment */
+    while (iter->state != ESTABLISHED && iter->state != CLOSE_WAIT) {
+        msg_receive(&msg);
+        switch (msg.type) {
+            case MSG_TYPE_NOTIFY_USER:
+                DEBUG("gnrc_conn_tcp : conn_tcp_accept() : MSG_TYPE_NOTIFY_USER\n");
+                break;
+
+            default:
+                DEBUG("gnrc_conn_tcp : conn_tcp_accept() : other message type\n");
+        }
+    }
+    return 0;
+}
+
+int conn_tcp_recv(conn_tcp_t *conn, void *data, size_t max_len)
+{
+    msg_t msg;
+    msg_t msg_tout;
+    xtimer_t usr_tout;
+    int err = 0;
+
+    assert(conn->l4_type == GNRC_NETTYPE_TCP);
+
+    /* Caller is not the owner of this connection */
+    if (conn->owner != thread_getpid()) {
+        return -EPERM;
+    }
+
+    /* If the connection is passive and the handle is not accepted */
+    if (conn->status & GNRC_TCP_STATUS_PASSIVE && !(conn->status & GNRC_TCP_STATUS_ACCEPTED)) {
+        return -EBADF;
+    }
+
+    /* Connection is in illegal State */
+    if (conn->state != ESTABLISHED && conn->state != FIN_WAIT_1
+        && conn->state != FIN_WAIT_2 && conn->state != CLOSE_WAIT) {
+        return -EALREADY;
+    }
+
```

Comment:
All `-EPERM`, `-EBADF` and `-EALREADY` could (and maybe should) be thrown from the state machine since this is essential to the state process itself.